ecl/hql/hqlopt.cpp

```diff
@@ -4269,9 +4269,29 @@ IHqlExpression * CTreeOptimizer::doCreateTransformed(IHqlExpression * transforme
 
                         if (ok)
                         {
-                            //If expanding the project removed all references to left (very silly join....) make it an all join
-                            if (transformed->hasAttribute(lookupAtom) && !exprReferencesDataset(&args.item(2), newLeft))
+                            //If expanding the project removed all references to left (very silly join....)
+                            //then it can be converted to a an all join against a filtered right dataset
+                            IHqlExpression * transformedCond = &args.item(2);
+                            if (transformed->hasAttribute(lookupAtom) && !exprReferencesDataset(transformedCond, newLeft))
+                            {
+                                IHqlExpression * rightDs = &args.item(1);
+                                OwnedHqlExpr right = createSelector(no_right, rightDs, transformedSeq);
+                                if (exprReferencesDataset(transformedCond, right))
+                                {
+                                    OwnedHqlExpr newCond = replaceSelector(transformedCond, right, rightDs);
+                                    OwnedHqlExpr filteredRightDs = createDataset(no_filter, { LINK(rightDs), LINK(newCond) });
+                                    //Replace the right dataset with the filtered dataset, and the join condition is now unconditional
+                                    args.replace(*filteredRightDs.getClear(), 1);
+                                    args.replace(*createConstant(true), 2);
+                                }
+                                if (transformed->hasAttribute(lookupAtom))
+                                {
+                                    if (!transformed->hasAttribute(manyAtom))
+                                        args.append(*createAttribute(keepAtom, createConstant(1)));
+                                    removeAttribute(args, lookupAtom);
+                                }
                                 args.append(*createAttribute(allAtom));
```

Comment:
I don't think this should apply if the join was originally local.

Reply:
JOIN(ds1, ds2, cond, ALL, LOCAL) is also supported - it does not distribute the right hand side.